File: java/datastructure/queue/priorityqueue/TestPriorityQueue.java
// 23. 合并K个升序链表 https://leetcode.cn/problems/merge-k-sorted-lists/
package datastructure.queue.priorityqueue;

/**
 * @Author john
 * @Description
 * @Date 2022/10/1 下午6:01
 */
public class TestPriorityQueue {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {}

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode mergeKLists(ListNode[] lists, Less<ListNode> cmp) {
        if (lists.length == 0) {
            return null;
        }

        PriorityQueue<ListNode> pq = new PriorityQueue<>(lists.length, cmp);
        for (ListNode node : lists) {
            if (node != null) {
                pq.push(node);
            }
        }

        ListNode dummy = new ListNode(-1), p = dummy;
        while (!pq.isEmpty()) {
            ListNode top = pq.pop();
            if (top.next != null) {
                pq.push(top.next);
            }
            p.next = top;
            p = p.next;
        }
        return dummy.next;
    }

    private ListNode[] createLists(int[][] arr) {
        ListNode[] lists = new ListNode[3];
        int i = 0;
        for (int[] a : arr) {
            ListNode p = new ListNode(-1), t = p;
            for (int v : a) {
                t.next = new ListNode(v);
                t = t.next;
            }
            lists[i++] = p.next;
        }
        return lists;
    }

    private void testPq(int[][] arr, Less<ListNode> cmp) {
        ListNode[] lists = createLists(arr);
        for (ListNode l : lists) {
            while (l != null) {
                System.out.printf("%d,", l.val);
                l = l.next;
            }
            System.out.println("");
        }
        ListNode p = mergeKLists(lists, cmp);
        while (p != null) {
            System.out.printf("%d,", p.val);
            p = p.next;
        }
        System.out.printf("\n\n");
    }

    public static void main(String[] args) {
        // [[1,4,5],[1,3,4],[2,6]]
        TestPriorityQueue tpq = new TestPriorityQueue();
        tpq.testPq(new int[][] {{1, 4, 5}, {1, 3, 4}, {2, 6}}, (o1, o2) -> o1.val < o2.val);
        tpq.testPq(new int[][] {{5, 4, 1}, {4, 3, 1}, {6, 2}}, (o1, o2) -> o1.val > o2.val);
    }
}

File: c/datastructure/queue/test/test_priority_queue.c
//
// Created by john on 2022/10/1.
//

#include "../priorityqueue/priorityqueue.h"
#include "../../linkedlist/linkedlist.h"

bool ListNodeLess(void *a, void *b) {
    return ((ListNode *) a)->val < ((ListNode *) b)->val;
}

ListNode *mergeKLists(ListNode **lists, int listsSize) {
    if (listsSize == 0) {
        return NULL;
    }
    PriorityQueue pq = createPriorityQueue(listsSize, ListNodeLess);
    for (int i = 0; i < listsSize; i++) {
        ListNode *l = lists[i];
        if (l) {
            push(&pq, l);
        }
    }
    ListNode *dummy = (ListNode *) malloc(sizeof(ListNode));
    dummy->next = NULL;
    ListNode *p = dummy;
    while (!isEmpty(&pq)) {
        ListNode *top = (ListNode *) pop(&pq);
        if (top->next) {
            push(&pq, top->next);
        }
        p->next = top;
        p = p->next;
    }
    releasePriorityQueue(&pq);
    p = dummy->next;
    free(dummy);
    return p;
}

LinkedList *createList(int *a, int n) {
    LinkedList *list = linkedListCreate();
    for (int i = 0; i < n; i++) {
        printf("%d,", a[i]);
        linkedListAddAtTail(list, a[i]);
    }
    printf("\n");
    return list;
}

// [[1,4,5],[1,3,4],[2,6]]
void test_priority_queue() {
    int a1[3] = {1, 4, 5}, a2[3] = {1, 3, 4}, a3[2] = {2, 6};
    LinkedList *l1 = createList(a1, 3), *l2 = createList(a2, 3), *l3 = createList(a3, 2);
    ListNode **lists = (ListNode **) malloc(3 * sizeof(ListNode *));
    lists[0] = l1->head->next;
    lists[1] = l2->head->next;
    lists[2] = l3->head->next;
    ListNode *t = mergeKLists(lists, 3), *p = t;
    while (p) {
        printf("%d,", p->val);
        ListNode *tmp = p;
        p = p->next;
        free(tmp);
    }
    free(lists);
    linkedListFree(l1);
    linkedListFree(l2);
    linkedListFree(l3);
}